File: game_server/tank.py
# game_server/tank.py
# Этот модуль определяет класс Tank, представляющий игровой танк
# с его свойствами и методами.
import time # Для временных меток событий
from core.message_broker_clients import send_kafka_message, KAFKA_DEFAULT_TOPIC_TANK_COORDINATES, KAFKA_DEFAULT_TOPIC_GAME_EVENTS
import logging # Добавляем логирование

logger = logging.getLogger(__name__) # Инициализация логгера для этого модуля

class Tank:
    """
    Представляет игровой танк.

    Содержит информацию о ID танка, его позиции, здоровье, статусе активности
    и времени последнего обновления. Предоставляет методы для движения,
    стрельбы, получения урона, сброса состояния и получения текущего состояния.
    Взаимодействует с Kafka для отправки событий.

    Атрибуты:
        tank_id (any): Уникальный идентификатор танка.
        position (tuple): Текущие координаты танка (например, (x, y)).
        health (int): Текущий уровень здоровья танка.
        is_active (bool): Флаг, указывающий, активен ли танк в данный момент (взят из пула).
        last_update_time (float): Время последнего обновления состояния танка (для дельта-обновлений).
    """
    def __init__(self, tank_id, initial_position=(0, 0), health=100):
        """
        Инициализирует объект танка.

        Args:
            tank_id (any): Уникальный ID для этого танка.
            initial_position (tuple, optional): Начальная позиция танка. По умолчанию (0, 0).
            health (int, optional): Начальное количество здоровья. По умолчанию 100.
        """
        self.tank_id = tank_id
        self.position = initial_position
        self.health = health
        self.is_active = False # Изначально танк не активен, пока не взят из пула
        self.last_update_time = 0 # Время последнего обновления (полезно для дельта-сжатия или интерполяции)
        logger.info(f"Танк {self.tank_id} создан с позицией {self.position} и здоровьем {self.health}.")

    def move(self, new_position):
        """
        Перемещает танк в новую позицию.

        Обновляет позицию танка и отправляет событие 'tank_moved' в Kafka.

        Args:
            new_position (tuple): Новые координаты (x, y) для танка.
        """
        self.position = new_position
        # Здесь может быть более сложная логика движения, валидация перемещения,
        # проверка столкновений и т.д.
        logger.info(f"Танк {self.tank_id} перемещен в {self.position}")
        # Отправка сообщения в Kafka о движении танка
        kafka_message = {
            "event_type": "tank_moved", # Тип события
            "tank_id": self.tank_id,    # ID танка
            "position": self.position,  # Новая позиция
            "timestamp": time.time()    # Временная метка события
        }
        send_kafka_message(KAFKA_DEFAULT_TOPIC_TANK_COORDINATES, kafka_message)

    def shoot(self):
        """
        Выполняет выстрел из танка.

        Логирует действие выстрела и отправляет событие 'tank_shot' в Kafka.
        Конкретная логика попадания, урона и т.д. здесь не реализована.
        """
        # Логика выстрела: создание снаряда, проверка попадания и т.д.
        logger.info(f"Танк {self.tank_id} стреляет из позиции {self.position}!")
        # Отправка сообщения в Kafka о выстреле
        kafka_message = {
            "event_type": "tank_shot",
            "tank_id": self.tank_id,
            "position_at_shot": self.position, # Позиция танка в момент выстрела
            "timestamp": time.time()
        }
        send_kafka_message(KAFKA_DEFAULT_TOPIC_GAME_EVENTS, kafka_message)
        # pass здесь означает, что дальнейшая логика выстрела (например, создание объекта снаряда)
        # может быть добавлена позже или обрабатывается другими системами.
        pass

    def take_damage(self, damage):
        """
        Уменьшает здоровье танка на указанное количество урона.

        Отправляет события 'tank_took_damage' и, если здоровье <= 0, 'tank_destroyed' в Kafka.
        Здоровье не может быть меньше 0.

        Args:
            damage (int): Количество полученного урона.
        """
        health_before_damage = self.health # Запоминаем здоровье до получения урона для лога
        self.health -= damage
        
        # Отправляем сообщение о получении урона
        kafka_damage_message = {
            "event_type": "tank_took_damage",
            "tank_id": self.tank_id,
            "damage_taken": damage,
            "current_health": max(0, self.health), # Логируем здоровье, ограниченное снизу нулем
            "health_before_damage": health_before_damage,
            "timestamp": time.time()
        }
        send_kafka_message(KAFKA_DEFAULT_TOPIC_GAME_EVENTS, kafka_damage_message)

        if self.health <= 0:
            self.health = 0 # Здоровье не может быть отрицательным
            logger.info(f"Танк {self.tank_id} уничтожен!")
            # Отправляем сообщение об уничтожении танка
            kafka_destroyed_message = {
                "event_type": "tank_destroyed",
                "tank_id": self.tank_id,
                "timestamp": time.time()
            }
            send_kafka_message(KAFKA_DEFAULT_TOPIC_GAME_EVENTS, kafka_destroyed_message)
            # Здесь может быть логика деактивации танка, удаления из игры и т.д.
        else:
            logger.info(f"Танк {self.tank_id} получил {damage} урона, текущее здоровье: {self.health}")

    def reset(self, initial_position=(0,0), health=100):
        """
        Сбрасывает состояние танка к начальным значениям.

        Используется при возвращении танка в пул объектов.

        Args:
            initial_position (tuple, optional): Начальная позиция для сброса. По умолчанию (0,0).
            health (int, optional): Начальное здоровье для сброса. По умолчанию 100.
        """
        self.position = initial_position
        self.health = health
        self.is_active = False # Помечаем как неактивный
        self.last_update_time = 0 # Сбрасываем время обновления
        logger.info(f"Танк {self.tank_id} был сброшен в начальное состояние.")

    def get_state(self):
        """
        Возвращает текущее состояние танка.

        Это состояние может использоваться для отправки клиентам или для логирования.

        Returns:
            dict: Словарь с состоянием танка (ID, позиция, здоровье).
        """
        return {
            'id': self.tank_id,
            'position': self.position,
            'health': self.health
            # Можно добавить и другие параметры, например, self.is_active, если это нужно клиенту
        }

File: cpp/game_server_cpp/tank.cpp
#include "tank.h"
#include <iostream> // Для std::cout, std::cerr для логирования
#include <ctime>    // Для std::time для временных меток

// Инициализация статических const членов
// Изменено KAFKA_TOPIC_TANK_COORDINATES на "tank_coordinates_history" согласно заданию
const std::string Tank::KAFKA_TOPIC_TANK_COORDINATES = "tank_coordinates_history";
const std::string Tank::KAFKA_TOPIC_GAME_EVENTS = "game_events";

Tank::Tank(std::string id,
           KafkaProducerHandler* kafka_handler,
           nlohmann::json initial_position,
           int initial_health)
    : tank_id_(std::move(id)),
      kafka_producer_handler_(kafka_handler),
      position_(std::move(initial_position)),
      health_(initial_health),
      is_active_(false) { // По умолчанию неактивен
    std::cout << "Tank " << tank_id_ << " created. Initial state: " << get_state().dump() << std::endl;
    if (!kafka_producer_handler_ || !kafka_producer_handler_->is_valid()) {
        std::cerr << "Tank " << tank_id_ << " Warning: KafkaProducerHandler is null or invalid. Kafka messages will not be sent." << std::endl;
    }
}

void Tank::move(const nlohmann::json& new_position) {
    if (!is_active_) {
        std::cout << "Tank " << tank_id_ << " is inactive. Move command ignored." << std::endl;
        return;
    }

    if (!new_position.contains("x") || !new_position.contains("y") ||
        !new_position["x"].is_number() || !new_position["y"].is_number()) {
        std::cerr << "Tank " << tank_id_ << " move error: new_position format is invalid. Expected {\"x\": number, \"y\": number}." << std::endl;
        return;
    }

    this->position_ = new_position;
    // std::cout << "Tank " << tank_id_ << " moved to " << this->position_.dump() << std::endl; // Может быть слишком подробно

    if (kafka_producer_handler_ && kafka_producer_handler_->is_valid()) {
        nlohmann::json kafka_message = {
            {"event_type", "tank_moved"},
            {"tank_id", this->tank_id_},
            {"position", this->position_}, // Изменено с "new_position" на "position" для согласованности
            {"timestamp", std::time(nullptr)}
        };
        kafka_producer_handler_->send_message(KAFKA_TOPIC_TANK_COORDINATES, kafka_message);
    }
}

void Tank::shoot() {
    if (!is_active_) {
        std::cout << "Tank " << tank_id_ << " is inactive. Shoot command ignored." << std::endl;
        return;
    }
    // std::cout << "Tank " << tank_id_ << " shoots!" << std::endl; // Может быть слишком подробно

    if (kafka_producer_handler_ && kafka_producer_handler_->is_valid()) {
        nlohmann::json kafka_message = {
            {"event_type", "tank_shot"},
            {"tank_id", this->tank_id_},
            {"position_at_shot", this->position_},
            {"timestamp", std::time(nullptr)}
        };
        kafka_producer_handler_->send_message(KAFKA_TOPIC_GAME_EVENTS, kafka_message);
    }
}

void Tank::take_damage(int damage) {
    if (damage <= 0) return; // Нет урона или лечения через этот метод

    this->health_ -= damage;
    bool destroyed = false;
    if (this->health_ <= 0) {
        this->health_ = 0;
        destroyed = true;
        // std::cout << "Tank " << tank_id_ << " took " << damage << " damage. Health is now " << this->health_ << ". Tank Destroyed!" << std::endl;
    } else {
        // std::cout << "Tank " << tank_id_ << " took " << damage << " damage. Health is now " << this->health_ << std::endl;
    }

    if (kafka_producer_handler_ && kafka_producer_handler_->is_valid()) {
        nlohmann::json damage_event_message = {
            {"event_type", "tank_took_damage"},
            {"tank_id", this->tank_id_},
            {"damage_amount", damage},
            {"current_health", this->health_},
            {"is_destroyed", destroyed}, // Добавлен флаг
            {"timestamp", std::time(nullptr)}
        };
        kafka_producer_handler_->send_message(KAFKA_TOPIC_GAME_EVENTS, damage_event_message);

        if (destroyed) {
            // Событие "tank_destroyed" может быть избыточным, если "tank_took_damage" включает "is_destroyed: true"
            // Однако, специфическое событие может быть полезно для разных потребителей.
            nlohmann::json destroyed_event_message = {
                {"event_type", "tank_destroyed"},
                {"tank_id", this->tank_id_},
                {"last_position", this->position_},
                {"timestamp", std::time(nullptr)}
            };
            kafka_producer_handler_->send_message(KAFKA_TOPIC_GAME_EVENTS, destroyed_event_message);
            // Деактивация должна обрабатываться игровой логикой/TankPool при получении "tank_destroyed" или достижении здоровья 0.
            // Сам танк не деактивирует себя просто от получения урона. set_active(false) вызывается при сбросе.
        }
    }
}

void Tank::reset(nlohmann::json initial_position, int health) {
    this->position_ = std::move(initial_position);
    this->health_ = health;
    bool old_active_status = this->is_active_;

    // set_active(false) обработает отправку события деактивации, если он был активен
    set_active(false);

    // std::cout << "Tank " << tank_id_ << " has been reset. New state: " << get_state().dump() << std::endl;

    if (kafka_producer_handler_ && kafka_producer_handler_->is_valid()) {
        nlohmann::json kafka_message = {
            {"event_type", "tank_reset"},
            {"tank_id", this->tank_id_},
            {"new_state", this->get_state()}, // get_state() теперь включает 'active:false'
            {"timestamp", std::time(nullptr)}
        };
        kafka_producer_handler_->send_message(KAFKA_TOPIC_GAME_EVENTS, kafka_message);
    }
}

void Tank::set_active(bool active_status) {
    if (this->is_active_ == active_status) { // Продолжаем, только если статус действительно изменился
        return;
    }

    this->is_active_ = active_status;
    // std::cout << "Tank " << tank_id_ << " active status set to: " << this->is_active_ << std::endl;

    if (kafka_producer_handler_ && kafka_producer_handler_->is_valid()) {
        nlohmann::json kafka_message = {
            {"event_type", this->is_active_ ? "tank_activated" : "tank_deactivated"},
            {"tank_id", this->tank_id_},
            {"current_state", this->get_state()}, // Отправляем полное состояние при активации/деактивации
            {"timestamp", std::time(nullptr)}
        };
        kafka_producer_handler_->send_message(KAFKA_TOPIC_GAME_EVENTS, kafka_message);
    }
}

nlohmann::json Tank::get_state() const {
    return {
        {"id", this->tank_id_},
        {"position", this->position_},
        {"health", this->health_},
        {"active", this->is_active_} // Изменено с "is_active" на "active" для согласованности
    };
}

bool Tank::is_active() const {
    return this->is_active_;
}

const std::string& Tank::get_id() const {
    return this->tank_id_;
}
